Sources/NIO/EventLoopFuture.swift

```diff
@@ -815,6 +815,28 @@ extension EventLoopFuture {
     }
 }
 
+extension EventLoopFuture {
+    /// Returns a new `EventLoopFuture` that fires only when this `EventLoopFuture` and all the provided futures complete.
+    ///
+    /// The returned `EventLoopFuture` will fail with the first error encountered, as soon as this `EventLoopFuture` or any of the provided futures fail: otherwise, it will succeed only when all of them do.
+    ///
+    /// It then provides the combined result, resulting from folding this `EventLoopFuture` with all of the provided futures using the `combingingFunction`.
+    ///
+    /// - parameters:
+    ///     - futures: An array of `EventLoopFuture<U>` to wait for.
+    ///     - eventLoop: The `EventLoop` on which the new `EventLoopFuture` callbacks will fire.
+    ///     - with: The combining function used to produce partial results.
+    /// - returns: A new `EventLoopFuture`.
+    public func fold<U>(_ futures: [EventLoopFuture<U>], eventLoop: EventLoop, with combiningFunction: @escaping (EventLoopFuture<T>, EventLoopFuture<U>) -> EventLoopFuture<T>) -> EventLoopFuture<T> {
```

Comment:
Hrm, is this exactly what we want this method to be? This method is not far from being a trivial wrapper over `reduce`, and there's a limited amount of utility here. I appreciate we may have been talking past each other here, but I was thinking we'd have something a bit closer to the Haskell construct, where the signature of the `combiningFunction` is `(T, U) -> EventLoopFuture<T>`.

I have no particular objection to the function as written, but I suspect the construct is of limited utility (you already have the array of futures, you can already call `reduce` on them, all `fold` does is wrap that into a new `promise` you just allocated).

Does that make sense?

Reply:
Ah I totally agree now that I think about it, I think I missed that point `(T, U) -> EventLoopFuture<T>`. 
I will modify this to reflect what we want it to be.

Reply:
👍 Thanks